### PR TITLE
Fix Escape passthrough from TextBox to running agents

### LINKED ISSUE #9704: Escape key not passed through to embedded Claude Code session (can't cancel/interrupt agent turn)
### Environment
- cmux **0.64.22 (102)**
- macOS Darwin 25.5.0
- Claude Code **2.1.223**

### Summary
Inside a cmux-managed Claude Code session, **`Escape` never reaches the Claude Code TUI**. In Claude Code, `Escape` is the documented way to **interrupt/cancel the AI's current turn** (stop generation or an in-progress action). Pressing `Escape` inside the focused CC pane does nothing — the agent keeps running, no interrupt happens.

This looks like the same class of bug as the closed #4993 (`Ctrl+F` not passed through, couldn't force-stop background agents). That issue's closing comment warned: *"the class stays: any ctrl-chord CC reads off the raw tty is one handler away from being swallowed."* `Escape` is not a ctrl-chord but the same principle applies — it's a keystroke Claude Code reads directly off the raw pty, and something between the window and the pty appears to be claiming it first (app-level shortcut layer / a modal-dismiss or fullscreen-exit handler, etc.) without ever forwarding it down, and without actually performing whatever it claimed the key for either.

### Ruled out (so the report is precise)
- **Not an IME/input-source issue**: reproduces identically with a CJK IME (Google Japanese Input) active AND with a plain ASCII input source ("ABC") active. Not related to `event.characters`/`charactersIgnoringModifiers` translation bugs like #1790/#2197.
- **Not a stuck runtime state**: fully quitting cmux (`Cmd+Q`, not just closing the window) and relaunching does not fix it — the session resumes via `autoResumeAgentSessions` and Escape still doesn't interrupt the new process.
- **Not global**: `Escape` works normally in every other macOS app on the same machine (Finder dialogs, Chrome forms, etc.) — this is scoped specifically to cmux's embedded Claude Code pane.
- **Not a cmux config issue**: `~/.config/cmux/cmux.json` `shortcuts.bindings` is untouched/default; the published `cmux.schema.json` has no user-configurable binding for Escape/interrupt/cancel at all — this appears to be hardcoded app behavior, not something a user can work around via settings.

### Steps to reproduce
1. Run a Claude Code session inside a cmux pane.
2. Let Claude start a turn (generating text, or running a tool).
3. Press `Escape` to cancel/interrupt, as Claude Code's own UI hints.
4. **Actual:** nothing happens — the keystroke never reaches CC, the agent keeps running/generating.
5. **Expected:** `Escape` reaches the focused CC session and interrupts the current turn, same as it would running `claude` directly in Terminal.app/iTerm2.

### Impact
There is currently no way to cancel/interrupt an in-progress Claude Code turn from inside cmux. The only escape hatch is fully quitting and relaunching cmux, which is disruptive and doesn't even reliably help if the agent starts a new turn immediately after resume.

### Suggested fix
Per the #4993 fix and its own closing comment, audit Escape passthrough the same way `Ctrl+F` passthrough was fixed — whatever app-level handler is claiming `Escape` first should either (a) yield to the focused CC pane's raw-tty read when CC is the focused responder, or (b) itself be confirmed to be a no-op in the reproduction scenario above so the real culprit layer can be found.

### PR DESCRIPTION
**Pull request by austinywang**
Fixes #9704.

## Root cause

A physical Escape reaches Claude Code when Ghostty owns first responder. When TextBoxInputTextView owns first responder, it consumes Escape and calls TerminalPanel.handleTextBoxEscape(); that action previously only focused the terminal and armed second-Escape hiding, so no Escape reached the PTY.

## Fix

TerminalPanel remains the single TextBox Escape action owner. It resolves live Dock ownership before workspace fallback, aggregates the authoritative per-panel lifecycle with the existing precedence while excluding manual loading keys, and forwards Escape only when that lifecycle is running. Existing focus and two-Escape-hide behavior remains intact in every state.

## Regression coverage

- Red test-only commit: 70ca1abd39. The focused run failed because the running workspace agent received zero key events: https://github.com/manaflow-ai/cmux/actions/runs/31164251236
- Fix commit: e9b1519896.
- Post-main-merge head: abca2707e0.
- Focused fixed-head run passed all four workspace/Dock running, idle, and manual-activity cases: https://github.com/manaflow-ai/cmux/actions/runs/31166301708

## Validation

- Tagged non-launch Debug build passed before the final main merge: CMUX_SKIP_ZIG_BUILD=1 ./scripts/reload.sh --tag issue-9704-escape-passthrough
- Post-merge project normalization, test wiring, determinism, and diff checks passed.
- Swift warning-budget validation passed; changed files introduce zero Swift warnings.
- Required PR status: 8 pass, 0 pending, 0 fail; GitHub reports MERGEABLE.
- Canonical branch autoreview and cmux policy review are clean on abca2707e0.
- No local Xcode tests or XCUITests were run.

## Dogfood evidence

Runtime reproduction was completed locally before the fix. Cloud video capture was blocked because the loader joined the Switchboard Tailscale tailnet and was unreachable from the local Mac, so no CUA/video capture was possible.

## Localization audit

No user-facing strings changed, so no localization catalog or web message update is required.

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/9810"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1788684162&installation_model_id=21920&pr_number=9810&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F9810&signature=cc0b9716163735d9efe639adb0d96cc714987e12207b8d9a82833220493dbeaa"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Passes Escape from the TextBox to the terminal when a running, allowlisted agent turn is active (now including `campfire`), restoring terminal interrupts while preserving TextBox focus and two‑Escape hide. Previously Escape was consumed by the TextBox and never reached the PTY.

- Terminal behavior: `TerminalPanel.handleTextBoxEscape()` sends Escape to the PTY only if the Dock-first then workspace-fallback lifecycle is `.running`; otherwise it just refocuses the terminal and arms hide‑on‑next‑Escape.
- Lifecycle: aggregation is centralized in `AgentHibernationLifecycleState.aggregate` and `aggregateForTextBoxEscape`; the latter filters to built‑in allowed keys and excludes manual keys. `Workspace` and `DockSplitStore` expose `agentLifecycleStateForTextBoxEscape`. Allowed keys now include `campfire`.
- Tests: `TextBoxEscapePassthroughTests` cover workspace/Dock running (Claude Code and Campfire), idle, manual, unsupported keys, and the stabilized physical TextBox keyDown path.

<sup>Written for commit 09dfa0ef4b243cac3ceda028c56b4ef90adc9d92. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/9810?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Low Risk**
> Localized input routing change gated on existing lifecycle aggregation; no auth or persistence changes. Regression tests cover main edge cases.
> 
> **Overview**
> **Escape from the TextBox now reaches the terminal when an agent turn is running**, fixing cases where TextBox owned first responder and consumed Escape without sending it to the PTY (e.g. interrupting Claude Code).
> 
> `TerminalPanel.handleTextBoxEscape()` checks aggregated panel lifecycle via Dock-first then workspace fallback; when state is `.running`, it sends Escape through `sendNamedKeyResult` before the existing focus-terminal and two-Escape-hide logic. Idle, unknown, and manual loading activity still do not passthrough.
> 
> Lifecycle aggregation is centralized in `AgentHibernationLifecycleState.aggregate` (manual keys excluded, same precedence as before). `Workspace` and `DockSplitStore` both call it through `agentHibernationLifecycleState`.
> 
> New `TextBoxEscapePassthroughTests` cover workspace and Dock running/idle/manual scenarios plus focused TextBox keyDown routing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 547aaf4522ede24fc9ce0fbe092a34bd3c13621a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Pressing Escape in supported running agent sessions now passes the key through to the terminal.
  * Escape passthrough works in workspace and Dock panels while preserving drafts and returning focus to the terminal.

* **Bug Fixes**
  * Escape passthrough is disabled for idle, unsupported, or manually active sessions.

* **Tests**
  * Added coverage for Escape interactions across supported panel configurations.

### CODE REVIEW

**Comment by austinywang** [comment]
Resolved the CodeRabbit package-boundary suggestion without a code change. AgentHibernationLifecycleState and its status-key policy predate this PR and are already intentionally compiled into both the app and CLI targets. This PR adds no new consumer or dependency edge; it centralizes the existing Workspace precedence so Dock ownership uses that same authoritative rule. A standalone CmuxAgentLifecycle package would be a narrow logic slice rather than the whole-domain boundary required by skills/cmux-architecture/SKILL.md, while migrating the full agent-lifecycle domain is unrelated to the Escape regression. The canonical cmux policy review and branch autoreview are both clean on abca2707e0.

File: vendor/bonsplit [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 18f3f8bc2ac3019d0f30b41586f4785950af6de7
+Subproject commit 4da56abb6d64595970af2081726591889ca54982
```
[the remark this answers is not in the document]
Comment (austinywang) [reply vendor/bonsplit]: Addressed in the current HEAD: vendor/bonsplit is restored to API-compatible 18f3f8bc2ac3019d0f30b41586f4785950af6de7, which is reachable from the submodule origin/main and contains the TabDragTransferRegistry/Registration APIs used by cmux. The older 4da56abb pointer was an intermediate base-sync artifact and is no longer present.

*— Claude Code*

File: ghostty [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit b17f1726fc53495dd827f7d85c9ec38da3d03814
+Subproject commit 11aa609d75dec882ef2f83171e2cbe887aeddbc5
```
[the remark this answers is not in the document]
Comment (austinywang) [reply ghostty]: Explicitly addressed and re-checked. Commit cc37e25080 restored the API-compatible Ghostty pin b17f1726fc. The integrated fix in PR #10959 intentionally leaves the current main pin at 5045df3f (the compatible main-side descendant); no API-removing downgrade remains. This PR is now superseded by #10959.

File: Sources/AgentHibernation/AgentHibernationLifecycleState.swift [hunk 3]
```diff
@@ -25,6 +25,23 @@ enum AgentHibernationLifecycleState: String, Codable, Sendable, Equatable, CaseI
         parse(rawValue)
     }
 
+    static func aggregate(
+        statusKeyedStates: [String: AgentHibernationLifecycleState],
+        fallback: AgentHibernationLifecycleState?
+    ) -> AgentHibernationLifecycleState {
+        let states = statusKeyedStates
+            .filter { !AgentHibernationLifecycleStatusKeys.isManualKey($0.key) }
+            .map(\.value)
```
[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/AgentHibernation/AgentHibernationLifecycleState.swift:34]: Explicitly addressed and re-checked. Commit 2b8847f4fd added the built-in-key allowlist path and unsupported-running coverage; merged PR #10959 carries the bounded allowlist implementation in 87f31977237cbcbbf8b7f492718685d612fbb9b0. Current main therefore fails closed for unsupported keys. This PR is superseded by #10959.

**Comment by austinywang** [comment]
## Final review audit (re-checked against HEAD)

Re-checked on 2026-09-01 against PR HEAD `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` and the current repository state.

The exact PR HEAD passed hosted run [33494544906](https://github.com/manaflow-ai/cmux/actions/runs/33494544906): `cmuxTests/TextBoxEscapePassthroughTests`, all 7 Swift Testing cases passed, and the workflow selector guard counted 7 executed tests. Current `main` also passed the same 7-case suite in [33498052518](https://github.com/manaflow-ai/cmux/actions/runs/33498052518). `main` already contains the canonical behavior through merged [PR #10959](https://github.com/manaflow-ai/cmux/pull/10959), commit `87f31977237cbcbbf8b7f492718685d612fbb9b0`, including the bounded allowlist implementation.

Because #9810 is now a conflicting duplicate of that merged change and carries an unrelated Ghostty submodule pointer, the expert disposition is to close #9810 as superseded rather than merge it. This is intentional; no additional code fix is required. No `CHANGES_REQUESTED` review is present. Every inline review thread has an explicit reply, including the two threads that had previously been auto-resolved without an author response.

### Top-level review bodies

- Codex: no GitHub top-level Codex body was posted; the prior structured Codex review found no actionable defect. There is no unanswered Codex ask.
- CodeRabbit review `5013590105`: its three actionable asks are covered in the table below and in the explicit thread replies.
- Greptile comment `5402942088`: its safe-to-merge summary was re-checked against both hosted runs and the merged integration; no new ask.
- Cubic summary in the PR description: the behavior is present in merged #10959 and covered by the 7-case hosted suite.
- Vercel and Cursor comments are deployment/spend-limit notices, not code requests.

### Comment audit

| Comment ID | Author | File:line | Ask | Disposition | Commit SHA |
|---|---|---|---|---|---|
| 3848349743 | coderabbitai[bot] | `ghostty:1` | Retain the API-compatible Ghostty pin | already-fixed | `cc37e2508017c2728eb2028c02c6a34290b11122`; canonical main fix `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| 3903144075 | austinywang | `ghostty:1` | Explicit reply to the Ghostty ask | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 3903144763 | coderabbitai[bot] | `ghostty:1` | Confirmation of the Ghostty resolution | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 3848349751 | coderabbitai[bot] | `Sources/AgentHibernation/AgentHibernationLifecycleState.swift:34` | Reject unsupported running keys and cover the case | already-fixed | `2b8847f4fd5fafb8c25282e92a0a16246e05c657`; canonical bounded path `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| 3903145226 | austinywang | `Sources/AgentHibernation/AgentHibernationLifecycleState.swift:34` | Explicit reply explaining the dedicated Escape allowlist | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 3903145788 | coderabbitai[bot] | `Sources/AgentHibernation/AgentHibernationLifecycleState.swift:34` | Confirmation of the lifecycle resolution | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 3848349761 | coderabbitai[bot] | `vendor/bonsplit:1` | Preserve the drag-transfer APIs | already-fixed | `3beb5321754c9d67f2f60d43bafb5df6ddc5a177`; API-compatible main baseline |
| 3848726530 | austinywang | `vendor/bonsplit:1` | Explicit Bonsplit response | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 3848727958 | coderabbitai[bot] | `vendor/bonsplit:1` | Confirmation of the Bonsplit resolution | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 5013590105 | coderabbitai[bot] | `ghostty`, lifecycle state, Bonsplit | Three actionable review asks | already-fixed | `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| 5214825398 | coderabbitai[bot] | `PR-level` | Review-paused notification; no code ask | disagree | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 5215343336 | coderabbitai[bot] | `PR-level` | Full-comparison notification; no code ask | disagree | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 5215397147 | austinywang | `PR-level` | Deliberate response rejecting a narrow package extraction | disagree | `7391d871d519a191c8e47a833f9aa317e7745050` |
| 5402924485 | coderabbitai[bot] | `PR-level` | Full-comparison notification; no code ask | disagree | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 5402942088 | greptile-apps[bot] | `PR-level` | Safe-to-merge summary; no blocking ask | already-fixed | `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| 5403016490 | cursor[bot] | `PR-level` | Spend-limit notification; no code ask | disagree | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| 5492162766 | vercel[bot] | `PR-level` | Deployment-status notification; no code ask | disagree | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |
| PR body (Cubic) | cubic | `PR description` | Generated behavior summary | already-fixed | `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| PR body (CodeRabbit) | coderabbitai[bot] | `PR description` | Generated release-note summary | already-fixed | `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| PR body (Cursor) | cursor[bot] | `PR description` | Generated risk summary | already-fixed | `87f31977237cbcbbf8b7f492718685d612fbb9b0` |
| Codex (no GitHub comment) | Codex | `PR-level` | No top-level body was published | already-fixed | `09dfa0ef4b243cac3ceda028c56b4ef90adc9d92` |

The three historical inline threads are all resolved and all have explicit replies; no review thread is being resolved silently.
